File: python/chapter_3/03_plates.py
# Как известно, слишком высокая стопка тарелок может развалиться.
# Следовательно, в реальной жизни, когда высота стопки превысила бы некоторое
# пороговое значение, мы начали бы складывать тарелки в новую стопку.
# Реализуйте структуру данных SetOfStacks, имитирующую реальную ситуацию.
# Структура SetOfStack должна состоять из нескольких стеков, новый стек
# создается, как только предыдущий достигнет порогового значения. Методы
# SetOfStacks.push() и SetOfStacks.pop() должны вести себя так же, как при
# работе с одним стеком (то есть метод pop() должен возвращать те же значения,
# которые бы он возвращал при использовании одного большого стека).
# Дополнительно:
# Реализуйте функцию popAt(int index), которая осуществляет операцию pop
# с заданным внутренним стеком.
import unittest


class Node:
    def __init__(self, value):
        self.value = value
        self.above = None
        self.below = None


class Stack:
    def __init__(self, capacity):
        self.size = 0
        self.top = None
        self.bottom = None
        self.capacity = capacity

    def __iter__(self):
        n = self.top
        while n:
            v = n.value
            n = n.below
            yield v

    def __str__(self):
        return ','.join([str(i) for i in self])

    def push(self, x):
        new_node = Node(x)
        self.size += 1
        if self.size == 1:
            self.bottom = self.top = new_node
            return
        self.top.above = new_node
        new_node.below = self.top
        self.top = new_node

    def pop(self):
        if not self.top:
            return None
        n = self.top
        self.top = self.top.below
        self.size -= 1
        return n.value

    def from_bottom(self):
        n = self.bottom
        self.bottom = self.bottom.above
        if self.bottom:
            self.bottom.below = None
        self.size -= 1
        return n.value

    def is_full(self):
        return self.size == self.capacity


class StackPlates:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stacks = []

    def push(self, x):
        stack = self.get_last_stack()
        if stack and not stack.is_full():
            stack.push(x)
        else:
            new_stack = Stack(self.capacity)
            new_stack.push(x)
            self.stacks.append(new_stack)

    def pop(self):
        stack = self.get_last_stack()
        if not stack:
            return None
        v = stack.pop()
        if stack.size == 0:
            del self.stacks[-1]
        return v

    def pop_at(self, index):
        return self.shift(index, True)

    def shift(self, index, from_top):
        stack = self.stacks[index]
        item = stack.pop() if from_top else stack.from_bottom()
        if stack.size == 0:
            del self.stacks[index]
        elif len(self.stacks) > index + 1:
            v = self.shift(index + 1, False)
            stack.push(v)
        return item

    def get_last_stack(self):
        if not self.stacks:
            return None
        return self.stacks[-1]


class Test(unittest.TestCase):
    s = StackPlates(5)

    def test_01_StackPlates(self):
        for i in range(15):
            self.s.push(i)
        actual = []
        for _ in range(15):
            actual.append(self.s.pop())
        self.assertEqual(list(reversed(range(15))), actual)

    def test_02_pop_at(self):
        for i in range(15):
            self.s.push(i)
        actual = []
        for _ in range(11):
            actual.append(self.s.pop_at(0))
        self.assertEqual(list(range(4, 15)), actual)


if __name__ == '__main__':
    unittest.main()

File: go/chapter_3/03_plates.go
package chapter_3

import "fmt"

type Plate struct {
	index int
	above *Plate
	below *Plate
}

type Pile struct {
	size     int
	capacity int
	top      *Plate
	bottom   *Plate
}

type RowOfPiles struct {
	capacity int
	piles    []*Pile
}

func GetPlate(index int) *Plate {
	return &Plate{index: index}
}

func GetPile(capacity int) *Pile {
	return &Pile{capacity: capacity}
}

func (p *Pile) push(value int) {
	newPlate := GetPlate(value)
	p.size++
	if p.size == 1 {
		p.bottom, p.top = newPlate, newPlate
		return
	}
	p.top.above = newPlate
	newPlate.below = p.top
	p.top = newPlate
}

func (p *Pile) pop() (int, error) {
	if p.top == nil {
		return -1, fmt.Errorf("Cannot pop. Pile is empty.")
	}
	n := p.top
	p.top = p.top.below
	p.size--
	return n.index, nil
}

func (p *Pile) fromBottom() int {
	n := p.bottom
	p.bottom = p.bottom.above
	if p.bottom != nil {
		p.bottom.below = nil
	}
	p.size--
	return n.index
}

func (p Pile) isFull() bool {
	return p.size == p.capacity
}

func GetRowOfPiles(capacity int) *RowOfPiles {
	return &RowOfPiles{capacity: capacity}
}

func (r *RowOfPiles) push(value int) {
	p := r.getLast()
	if p != nil && !p.isFull() {
		p.push(value)
	} else {
		newPile := GetPile(r.capacity)
		newPile.push(value)
		r.piles = append(r.piles, newPile)
	}
}

func (r *RowOfPiles) pop() (int, error) {
	p := r.getLast()
	if p == nil {
		return -1, fmt.Errorf("Cannot pop. There aren't piles.")
	}
	ret, err := p.pop()
	if err != nil {
		return -1, err
	}
	if p.size == 0 {
		r.piles = r.piles[:len(r.piles)-1]
	}
	return ret, nil
}

func (r *RowOfPiles) popAt(pileIndex int) int {
	return r.shift(pileIndex, true)
}

func (r *RowOfPiles) shift(pileIndex int, fromTop bool) int {
	var ret int

	p := r.piles[pileIndex-1]
	if fromTop {
		ret, _ = p.pop()
	} else {
		ret = p.fromBottom()
	}

	if p.size == 0 {
		r.piles = r.piles[:len(r.piles)-1]
	} else if len(r.piles) > pileIndex {
		tmp := r.shift(pileIndex+1, false)
		p.push(tmp)
	}
	return ret
}

func (r RowOfPiles) getLast() *Pile {
	if len(r.piles) == 0 {
		return nil
	}
	return r.piles[len(r.piles)-1]
}
